applications/nrf_desktop/src/hw_interface/click_detector.c

```diff
@@ -0,0 +1,265 @@
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CLICK_DETECTOR_LOG_LEVEL);
+
+#define CLICK_CHECK_PERIOD	K_MSEC(100)
+#define CLICK_ON_INIT_MAX	K_MSEC(500)
+
+#define SHORT_CLICK_MAX		K_MSEC(500)
+#define LONG_CLICK_MIN		K_MSEC(5000)
+
+#define CLICK_HOLD_CHECK_CNT_MAX (LONG_CLICK_MIN/CLICK_CHECK_PERIOD)
+
+enum state {
+	STATE_DISABLED,
+	STATE_INIT_CLICK_CHECK,
+	STATE_INIT_HOLD_CHECK,
+	STATE_ACTIVE,
+	STATE_SLEEP,
+};
+
+struct key_state {
+	bool pressed_on_init;
+	bool pressed;
+	int click_short_timeout;
+	size_t pressed_time;
+};
+
+struct key_state keys[ARRAY_SIZE(key_configs)];
+
+static enum state state;
+struct k_delayed_work click_check;
+
+
+static void submit_click_event(u16_t key_id, enum click click)
+{
+	struct click_event *event = new_click_event();
+
+	event->key_id = key_id;
+	event->click = click;
+
+	EVENT_SUBMIT(event);
+}
+
+static void click_check_fn(struct k_work *work)
+{
+	BUILD_ASSERT(CLICK_HOLD_CHECK_CNT_MAX <= UCHAR_MAX);
+	static u8_t init_hold_check_cnt;
+
+	switch (state) {
+	case STATE_INIT_CLICK_CHECK:
+	{
+		bool click_active = false;
+
+		for (size_t i = 0; i < ARRAY_SIZE(keys); i++) {
+			if (keys[i].pressed_on_init) {
+				click_active = true;
+				break;
+			}
+		}
+
+		if (click_active) {
+			state = STATE_INIT_HOLD_CHECK;
+		} else {
+			state = STATE_ACTIVE;
+		}
+		break;
+	}
+
+	case STATE_INIT_HOLD_CHECK:
+		init_hold_check_cnt++;
+		if (init_hold_check_cnt == CLICK_HOLD_CHECK_CNT_MAX) {
+			for (size_t i = 0; i < ARRAY_SIZE(keys); i++) {
+				if (keys[i].pressed_on_init) {
+					submit_click_event(key_configs[i].key_id,
+							   CLICK_HOLD_ON_INIT);
+				}
+			}
+			state = STATE_ACTIVE;
+		}
+		/* Fall through. */
+	case STATE_ACTIVE:
+		for (size_t i = 0; i < ARRAY_SIZE(keys); i++) {
+			if (keys[i].click_short_timeout) {
+				keys[i].click_short_timeout -= CLICK_CHECK_PERIOD;
+				if (keys[i].click_short_timeout <= 0) {
+					submit_click_event(key_configs[i].key_id,
+							   CLICK_SHORT);
+					keys[i].click_short_timeout = 0;
+				}
+			}
+			if (keys[i].pressed) {
+				keys[i].pressed_time += CLICK_CHECK_PERIOD;
+
+				if (keys[i].pressed_time > LONG_CLICK_MIN) {
+					keys[i].pressed = false;
+					keys[i].pressed_time = 0;
+					submit_click_event(key_configs[i].key_id,
+							   CLICK_LONG);
+				}
+			}
+		}
+		break;
+
+	default:
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+
+	k_delayed_work_submit(&click_check, CLICK_CHECK_PERIOD);
+}
+
+static int init(void)
+{
+	BUILD_ASSERT(ARRAY_SIZE(keys) == ARRAY_SIZE(key_configs));
+
+	k_delayed_work_init(&click_check, click_check_fn);
+	k_delayed_work_submit(&click_check, CLICK_ON_INIT_MAX);
+
+	state = STATE_INIT_CLICK_CHECK;
+
+	return 0;
+}
+
+static enum click get_click(u32_t time_diff)
+{
+	if (time_diff < SHORT_CLICK_MAX) {
+		return CLICK_SHORT;
+	}
+	return CLICK_NONE;
+}
+
+static void process_key(struct key_state *key, u16_t key_id, bool pressed)
+{
+	switch (state) {
+	case STATE_INIT_CLICK_CHECK:
+		key->pressed_on_init = pressed;
```

Comment:
let's keep the logic related to key held on init in the module that uses it

Reply:
ok